### PR TITLE
Don't enforce new validation rules for existing networks

### LINKED ISSUE #47331: Existing Docker Swarm Network with invalid subnet config is not upgraded gracefully when upgrading to 25.x
### Description

We have clusters that have a manually created ingress network created with a subnet 10.0.0.1/16. While this is not a valid subnet block, we never had any issues with it.

Starting with https://github.com/moby/moby/pull/46183 it seems like validation upon container creation was improved. This however inadvertently caused our clusters to break upon upgrade.

I know the proper fix would be to recreate the ingress with 10.0.0.0/16, but that might not be the best course of action for production clusters.

### Reproduce

1. Create Cluster on Docker 24.x
2. Remove ingress network `docker network rm ingress`
3. Create ingress network with broken subnet range `docker network create --driver overlay --ingress ingress --subnet=10.0.0.1/16
`
4. Create service using the network 
5. Upgrade cluster to 25.x
6. see that service is not being brought up anymore

### Expected behavior

Existing (overlay) networks should not stop working when upgrading Docker

### docker version

```bash
Client: Docker Engine - Community
 Version:           25.0.2
 API version:       1.44
 Go version:        go1.21.6
 Git commit:        29cf629
 Built:             Thu Feb  1 00:23:19 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          25.0.2
  API version:      1.44 (minimum version 1.24)
  Go version:       go1.21.6
  Git commit:       fce6e0c
  Built:            Thu Feb  1 00:23:19 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.28
  GitCommit:        ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client: Docker Engine - Community
 Version:    25.0.2
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.12.1
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.24.5
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
  scan: Docker Scan (Docker Inc.)
    Version:  v0.17.0
    Path:     /usr/libexec/docker/cli-plugins/docker-scan

Server:
 Containers: 57
  Running: 20
  Paused: 0
  Stopped: 37
 Images: 29
 Server Version: 25.0.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: syslog
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: active
  NodeID: c20vpmwul4vai14swgkpbj8jd
  Is Manager: true
  ClusterID: ljgyd4c405vfzvwq339oif3gt
  Managers: 3
  Nodes: 3
  Default Address Pool: 10.0.0.0/8  
  SubnetSize: 24
  Data Path Port: 4789
  Orchestration:
   Task History Retention Limit: 5
  Raft:
   Snapshot Interval: 10000
   Number of Old Snapshots to Retain: 0
   Heartbeat Tick: 1
   Election Tick: 10
  Dispatcher:
   Heartbeat Period: 5 seconds
  CA Configuration:
   Expiry Duration: 3 months
   Force Rotate: 0
  Autolock Managers: false
  Root Rotation In Progress: false
  Node Address: 192.168.0.4
  Manager Addresses:
   192.168.0.4:2377
   192.168.0.5:2377
   192.168.0.6:2377
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc version: v1.1.12-0-g51d5e94
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
 Kernel Version: 5.4.0-170-generic
 Operating System: Ubuntu 20.04.6 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 7.574GiB
 Name: precisionx-swarm-prod-01
 ID: PULU:MERA:NH2U:EGMA:ZC5P:AMYT:Y7UB:M6RI:7VJ6:Q5BE:EAMZ:UFLP
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Default Address Pools:
   Base: 172.18.0.0/16, Size: 24

WARNING: No swap limit support
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by robmry**
**- What I did**

Non-swarm networks created before validation checks were added on network creation (https://github.com/moby/moby/pull/45759) continued working in 25.0.x because the checks were not re-run.

But, a swarm network with invalid IPAM config (for example, subnet `10.0.0.1/24`) fails to initialise on upgrade to 25.0.x.

Fixes https://github.com/moby/moby/issues/47331

**- How I did it**

Swarm creates networks (with `agent=true`) to ensure they exist on each agent, when needed - ignoring the NetworkNameError that says the network already existed.

By ignoring IPAM validation errors on creation of a network with `agent=true`, pre-existing swarm networks with IPAM config that would fail the new checks will continue to work too.

New swarm (overlay) networks are still validated, because they are initially created with 'agent=false'.

**- How to verify it**

As described in https://github.com/moby/moby/issues/47331 ...

Start with a 24.0.x build:
- docker network rm ingress
- docker network create --driver overlay --ingress ingress --subnet=10.0.0.1/1
- docker service create -p 80:80 nginx
- Upgrade to docker 25.0.3

The service failed to start, logging errors like this ...

```
level=error msg="fatal task error" error="invalid network config:\ninvalid subnet 10.0.0.1/16: it should be 10.0.0.0/16" module=node/agent/taskmanager node.id=s6ui3u8t66gml6czf2jwxhpgn service.id=rty5ui8sw8j6a1yl5tl98e6z9 task.id=lffsjphhk89mcwhhvn6heb91x
```

With this change, the service starts, and log lines like this are generated ("ingress" is configured as above, "oln" is a non-ingress overlay network) ...

```
WARN[2024-02-08T17:38:29.881000427Z] Continuing with validation errors in agent network IPAM  error="invalid network config:\ninvalid subnet 10.0.0.1/16: it should be 10.0.0.0/16" network=ingress
WARN[2024-02-08T17:38:29.881073302Z] Continuing with validation errors in agent network IPAM  error="invalid network config:\ninvalid subnet 10.1.0.1/16: it should be 10.1.0.0/16" network=oln
```

It's not possible to create new overlay and bridge networks with the same issue ...

```
# docker network create --subnet 10.10.0.1/24 nnn
Error response from daemon: invalid network config:
invalid subnet 10.10.0.1/24: it should be 10.10.0.0/24

# docker network create --driver overlay ooo --subnet=10.2.0.1/16
Error response from daemon: invalid network config:
invalid subnet 10.2.0.1/16: it should be 10.2.0.0/16
```

**- Description for the changelog**

Do not enforce new validation rules for existing swarm networks

### CODE REVIEW

File: daemon/network.go [hunk 1]
```diff
@@ -332,7 +332,19 @@ func (daemon *Daemon) createNetwork(cfg *config.Config, create types.NetworkCrea
 	}
 
 	if err := network.ValidateIPAM(create.IPAM, create.EnableIPv6); err != nil {
-		return nil, errdefs.InvalidParameter(err)
+		if agent {
+			// Swarm creates networks with agent=true to ensure that they exist on each agent
+			// when needed. Non-swarm networks created before ValidateIPAM was introduced
+			// continue to work, because they're not seen again here. By ignoring errors for
+			// agent networks here, swarm networks will continue to work too. (New overlay
+			// networks are initially created as !agent, so they are still validated.)
```
Comment (corhere) [inline daemon/network.go:340]: When a swarm-scoped network is created it is run through `createNetwork` with `!agent`, but the network is not actually created by the `createNetwork` call. And this is true for all swarm-scoped networks irrespective of driver, not just overlay. (#32981)

**Review by corhere — COMMENTED** [review]
I now see how newly-created networks are always validated before passed to Swarmkit: the `(*Daemon).createNetwork(agent=false)` flow runs as usual, all the way up to the `(*Controller).NewNetwork()` call. Libnetwork notices that it's a Swarm-scope network and returns `libnetwork.ManagerRedirectError`, and the caller creates the network in the Swarm state instead.
https://github.com/moby/moby/blob/7cb1efebec36e43bcfc256f4558610c13f289b05/api/server/router/network/network_routes.go#L202-L231

In light of this I am okay with your fix as-is, under one condition: a comment is added in `func (*networkRouter) postNetworkCreate` which calls out that the `n.backend.CreateNetwork(create)` call is necessary for validating the network spec for Swarm-scoped networks, despite not actually creating the network. I'm just a bit worried that we might accidentally regress with an overzealous libnetwork refactoring, given how surprising the control flow is.

**Comment by robmry** [comment]
Thank you @corhere, I've made those updates.

**Review by corhere — APPROVED** [review]

**Review by akerouanton — APPROVED** [review]
LGTM

**Review by thaJeztah — APPROVED** [review]
LGTM

**Pull request merged**